### ISSUE TITLE
[core] Restore constraint lambda in case of integration failure.

### ISSUE THREAD

**Issue by duburcqa** [duburcqa/jiminy#725]
Currently, the lambda multipliers of the constraints are not reset to their original value if the integration step failed. This is an issue because they are used as initial guess for the constraint solver. This coupling seems to be problematic on `gym_jiminy.envs:atlas` environment.  Here is a snippet to reproduce the issue:
```
import numpy as np
import gymnasium as gym
env = gym.make("gym_jiminy.envs:atlas", debug=True)
env.reset()
env.simulator.step(0.73)
for _ in range(10):
    print("-----")
    env.simulator.step(1e-6)
```

An additional buffers with the values at the previous iteration should be stored next to `contactForcesPrev_`, `fPrev_` and `aPrev_`.

**Comment by duburcqa** [comment]
It appears that the root cause of the exception (`Too many successive constraint solving failures.`) is not related to integration failure but rather inherent numerical instabilities of the contact solver, which is known to have poor convergence property. Here is a figure of the joint acceleration before raising the exception:
<img width="1728" alt="image" src="https://github.com/duburcqa/jiminy/assets/17752950/66d616d7-599f-4ef3-b556-e0d479a0b86b">

Smoothness of the accelerations is a major concern, in terms of simulation speed but also learning speed.

**Comment by duburcqa** [comment]
Two options at may improve convergence:
* implement SOR (successive over-relaxation with rfactor going from 2.0 to 0.0 linearly)
* implement SSOR (symmetric SOR) with factor fixed to 1.0

**Comment by duburcqa** [comment]
Before implementing under-relaxation:

<img width="1728" alt="Screenshot 2024-02-17 at 14 00 27" src="https://github.com/duburcqa/jiminy/assets/17752950/a41b7562-9248-4db7-bf38-7b08c0af6c52">

After (https://github.com/duburcqa/jiminy/pull/726):

<img width="1728" alt="Screenshot 2024-02-17 at 14 03 08" src="https://github.com/duburcqa/jiminy/assets/17752950/9724e46b-4d58-4260-8259-796bfc428664">
